# Notes on Julia

## Install Julia on a local Host

Play-jupyter with Julia requires julia language to be installed
- installed julia`sudo pamac install julia` 
- installed julia kernel through Julia "REPL":

```
    julia> using Pkg
    julia> Pkg.add("IJulia")
    Installing known registries into `~/.julia`
    ...
    Building IJulia → `~/.julia/packages/IJulia/IDNmS/deps/build.log`
    julia> 
```

- you may install julia kernel non-interactive 
  by executing `julia -e 'using Pkg; Pkg.add("IJulia")'`.
  Don't know if this could be part of `pipfile`.

## Julia Cheatsheet

[Julia Cheat Sheet for v1.0](https://juliadocs.github.io/Julia-Cheat-Sheet/) 


In [1]:
Base.VERSION

v"1.9.4"

During my first look - and what I kept in mind - was a function to display all methods of a type(?) ...

In [2]:
methods(42)

# 0 methods for callable object

... no, display all methods of function instead.

Here is the documentation I found to recall:
https://docs.julialang.org/en/v1/manual/methods/#Defining-Methods

As an example, generic function '+' has got a lot of methods.

In [3]:
methods(+)

# 207 methods for generic function "+" from Base:
   [1] +(x::T, y::T) where T<:Union{Int128, Int16, Int32, Int64, Int8, UInt128, UInt16, UInt32, UInt64, UInt8}
     @ int.jl:87
   [2] +(x::T, y::T) where T<:Union{Float16, Float32, Float64}
     @ float.jl:408
   [3] +(c::Union{UInt16, UInt32, UInt64, UInt8}, x::BigInt)
     @ Base.GMP gmp.jl:539
   [4] +(c::Union{Int16, Int32, Int64, Int8}, x::BigInt)
     @ Base.GMP gmp.jl:545
   [5] +(c::Union{UInt16, UInt32, UInt64, UInt8}, x::BigFloat)
     @ Base.MPFR mpfr.jl:407
   [6] +(c::Union{Int16, Int32, Int64, Int8}, x::BigFloat)
     @ Base.MPFR mpfr.jl:415
   [7] +(c::Union{Float16, Float32, Float64}, x::BigFloat)
     @ Base.MPFR mpfr.jl:423
   [8] +(x::Union{Dates.CompoundPeriod, Dates.Period})
     @ Dates /usr/share/julia/stdlib/v1.9/Dates/src/periods.jl:342
   [9] +(x::Dates.Instant)
     @ Dates /usr/share/julia/stdlib/v1.9/Dates/src/arithmetic.jl:4
  [10] +(A::LinearAlgebra.UpperTriangular, B::LinearAlgebra.UpperTriangular)
     @ LinearAlgebra /usr/share/julia/stdlib/v1.9/LinearAlgebra/src/triangular.jl:642
  [11] +(A::LinearAlgebra.UpperTriangular, B::LinearAlgebra.UnitUpperTriangular)
     @ LinearAlgebra /usr/share/julia/stdlib/v1.9/LinearAlgebra/src/triangular.jl:644
  [12] +(x::LinearAlgebra.UpperTriangular, H::LinearAlgebra.UpperHessenberg)
     @ LinearAlgebra /usr/share/julia/stdlib/v1.9/LinearAlgebra/src/hessenberg.jl:112
  [13] +(A::LinearAlgebra.UpperTriangular, B::LinearAlgebra.Bidiagonal)
     @ LinearAlgebra /usr/share/julia/stdlib/v1.9/LinearAlgebra/src/special.jl:91
  [14] +(r1::StepRangeLen{T, R}, r2::StepRangeLen{T, R}) where {R<:Base.TwicePrecision, T}
     @ twiceprecision.jl:628
  [15] +(r1::StepRangeLen{T, S}, r2::StepRangeLen{T, S}) where {T, S}
     @ range.jl:1464
  [16] +(F::LinearAlgebra.Hessenberg, J::LinearAlgebra.UniformScaling)
     @ LinearAlgebra /usr/share/julia/stdlib/v1.9/LinearAlgebra/src/hessenberg.jl:647
  [17] +(y::AbstractFloat, x::Bool)
     @ bool.jl:176
  [18] +(Da::LinearAlgebra.Diagonal, Db::LinearAlgebra.Diagonal)
     @ LinearAlgebra /usr/share/julia/stdlib/v1.9/LinearAlgebra/src/diagonal.jl:184
  [19] +(D::LinearAlgebra.Diagonal, S::LinearAlgebra.Symmetric)
     @ LinearAlgebra /usr/share/julia/stdlib/v1.9/LinearAlgebra/src/diagonal.jl:188
  [20] +(D::LinearAlgebra.Diagonal{var"#s972", V} where {var"#s972"<:Real, V<:AbstractVector{var"#s972"}}, H::LinearAlgebra.Hermitian)
     @ LinearAlgebra /usr/share/julia/stdlib/v1.9/LinearAlgebra/src/diagonal.jl:194
  [21] +(x::LinearAlgebra.Diagonal, H::LinearAlgebra.UpperHessenberg)
     @ LinearAlgebra /usr/share/julia/stdlib/v1.9/LinearAlgebra/src/hessenberg.jl:112
  [22] +(A::LinearAlgebra.Diagonal, B::LinearAlgebra.Bidiagonal)
     @ LinearAlgebra /usr/share/julia/stdlib/v1.9/LinearAlgebra/src/special.jl:130
  [23] +(A::LinearAlgebra.Diagonal, B::LinearAlgebra.SymTridiagonal)
     @ LinearAlgebra /usr/share/julia/stdlib/v1.9/LinearAlgebra/src/special.jl:140
  [24] +(A::LinearAlgebra.Diagonal, B::LinearAlgebra.Tridiagonal)
     @ LinearAlgebra /usr/share/julia/stdlib/v1.9/LinearAlgebra/src/special.jl:168
  [25] +(A::LinearAlgebra.Diagonal{var"#s972", V} where {var"#s972"<:Number, V<:AbstractVector{var"#s972"}}, B::LinearAlgebra.UniformScaling)
     @ LinearAlgebra /usr/share/julia/stdlib/v1.9/LinearAlgebra/src/special.jl:246
  [26] +(x::AbstractIrrational, y::AbstractIrrational)
     @ irrationals.jl:161
  [27] +(A::LinearAlgebra.Tridiagonal, B::LinearAlgebra.Tridiagonal)
     @ LinearAlgebra /usr/share/julia/stdlib/v1.9/LinearAlgebra/src/tridiag.jl:754
  [28] +(x::LinearAlgebra.Tridiagonal, H::LinearAlgebra.UpperHessenberg)
     @ LinearAlgebra /usr/share/julia/stdlib/v1.9/LinearAlgebra/src/hessenberg.jl:112
  [29] +(A::LinearAlgebra.Tridiagonal, B::LinearAlgebra.SymTridiagonal)
     @ LinearAlgebra /usr/share/julia/stdlib/v1.9/LinearAlgebra/src/special.jl:162
  [30] +(A::LinearAlgebra.Tridiagonal, B::LinearAlgebra.Diagonal)
     @ LinearAlgebra /usr/share/julia/stdlib/v1.9/Lin

In [4]:
f(x::Float64, y::Float64) = 2x + y

f (generic function with 1 method)

In [5]:
f(2.0, 3.0)

7.0

In [6]:
f(2.0, 3)

LoadError: MethodError: no method matching f(::Float64, ::Int64)

Closest candidates are:
  f(::Float64, ::Float64)
   @ Main In[4]:1


In [7]:
f(x::Number, y::Number) = 2x - y

f (generic function with 2 methods)

In [8]:
f(2.0, 3)

1.0

In [9]:
methods(f)

# 2 methods for generic function "f" from Main:
 [1] f(x::Float64, y::Float64)
     @ In[4]:1
 [2] f(x::Number, y::Number)
     @ In[7]:1

# Update as of January 2022

After a long time I revisited "my" notebook server and found IJulia not working any longer (don't know why, may be an update of Julia environment)

```
    [I 20:24:49.362 NotebookApp] Kernel started: 80c94540-a495-4416-bf92-f5460e177e18, name: julia-1.5
    ERROR: LoadError: ArgumentError: Package IJulia not found in current path:
    - Run `import Pkg; Pkg.add("IJulia")` to install the IJulia package.
    ...
```


Right, package IJulia is not in environment

```
    julia -e 'using Pkg; Pkg.status()'
        Status `~/.julia/environments/v1.7/Project.toml` (empty project)
```


So let's reinstall it, IJulia for Julia has as many dependencies as EvCxR has for Rust.

```   
    julia -e 'using Pkg; Pkg.add("IJulia")'
        Updating registry at `~/.julia/registries/General`
      Resolving package versions...
      Installed Conda ────────── v1.6.0
      Installed ZeroMQ_jll ───── v4.3.4+0
      Installed VersionParsing ─ v1.3.0
      Installed Preferences ──── v1.2.3
      Installed Parsers ──────── v2.2.0
      Installed JSON ─────────── v0.21.2
      Installed IJulia ───────── v1.23.2
      Installed JLLWrappers ──── v1.4.0
      Installed libsodium_jll ── v1.0.20+0
      Downloaded artifact: ZeroMQ
      Downloaded artifact: libsodium
        Updating `~/.julia/environments/v1.7/Project.toml`
      [7073ff75] + IJulia v1.23.2
        Updating `~/.julia/environments/v1.7/Manifest.toml`
      [8f4d0f93] + Conda v1.6.0
      [7073ff75] + IJulia v1.23.2
      [692b3bcd] + JLLWrappers v1.4.0
      [682c06a0] + JSON v0.21.2
      [739be429] + MbedTLS v1.0.3
      [69de0a69] + Parsers v2.2.0
      [21216c6a] + Preferences v1.2.3
      [b85f4697] + SoftGlobalScope v1.1.0
      [81def892] + VersionParsing v1.3.0
      [c2297ded] + ZMQ v1.2.1
      [8f1865be] + ZeroMQ_jll v4.3.4+0
      [a9144af2] + libsodium_jll v1.0.20+0
      [0dad84c5] + ArgTools
      [56f22d72] + Artifacts
      [2a0f44e3] + Base64
      [ade2ca70] + Dates
      [f43a241f] + Downloads
      [b27032c2] + LibCURL
      [76f85450] + LibGit2
      [8f399da3] + Libdl
      [56ddb016] + Logging
      [d6f4376e] + Markdown
      [a63ad114] + Mmap
      [ca575930] + NetworkOptions
      [44cfe95a] + Pkg
      [de0858da] + Printf
      [3fa0cd96] + REPL
      [9a3f8284] + Random
      [ea8e919c] + SHA
      [9e88b42a] + Serialization
      [6462fe0b] + Sockets
      [fa267f1f] + TOML
      [a4e569a6] + Tar
      [8dfed614] + Test
      [cf7118a7] + UUIDs
      [4ec0a83e] + Unicode
      [deac9b47] + LibCURL_jll
      [29816b5a] + LibSSH2_jll
      [c8ffd9c3] + MbedTLS_jll
      [14a3606d] + MozillaCACerts_jll
      [83775a58] + Zlib_jll
      [8e850ede] + nghttp2_jll
      [3f19e933] + p7zip_jll
        Building Conda ─→ `~/.julia/scratchspaces/44cfe95a-1eb2-52ea-b672-e2afdf69b78f/6cdc8832ba11c7695f494c9d9a1c31e90959ce0f/build.log`
        Building IJulia → `~/.julia/scratchspaces/44cfe95a-1eb2-52ea-b672-e2afdf69b78f/d8b9c31196e1dd92181cd0f5760ca2d2ffb4ac0f/build.log`
    Precompiling project...
      11 dependencies successfully precompiled in 4 seconds (4 already precompiled)

    julia -e 'using Pkg; Pkg.status()'
          Status `~/.julia/environments/v1.7/Project.toml`
      [7073ff75] IJulia v1.23.2
```


# Update as of December 2023

Same procedure as every year ... after a long time I revisited "my" notebook server and found IJulia not working any longer, the system has updated Julia to version 1.9.

```   
[tilo@holm play-jupyter]$ julia -e 'using Pkg; Pkg.status()'
Status `~/.julia/environments/v1.9/Project.toml`
  [7073ff75] IJulia v1.24.2
[tilo@holm play-jupyter]$ 
```   
